Add `class` command

njxue opened the pull request:
1.  `class [TAG]` to enter focus mode. Will replace tags with classes later
     * Lists students with the specified `TAG`
     * Sets focus mode to active state
2. `FocusMode` class in `model` package keeps track of current focus mode

njxue committed:
Add filtering of contact list by tags

njxue committed:
Merge branch 'master' of https://github.com/AY2223S1-CS2103T-T12-1/tp into add-tags-predicate

njxue committed:
Allow tracking of focus mode state

Bubbl3T commented on src/main/java/seedu/address/model/ReadOnlyFocusMode.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @param classToFocus Class corresponding to the focus mode.
+     */
+    void setFocusedClass(Tag classToFocus);
```
Since the interface's intention is such that the focus mode is read only, should we still allow it to set the focused class?

Bubbl3T commented on src/main/java/seedu/address/model/FocusMode.java:
```diff
@@ -0,0 +1,54 @@
+     * Exits focus mode.
+     */
+    @Override
+    public void exit() {
+        this.isActive = false;
+    }
```
Should we consider clearing the content of the `focusedClass` variable to prevent possible bugs?

Bubbl3T reviewed (changes requested):
Minor nits found and possible bug prevention through minor edits. Looks good otherwise.

njxue committed:
Remove setter method in FocusMode and set focused class to null when inactive

RezwanArefin01 commented on src/main/java/seedu/address/model/FocusMode.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class FocusMode implements ReadOnlyFocusMode {
+
+    private boolean isActive;
```
Is the `isActive` variable necessary? I think it is recoverable from `focusedClass != null`. So this can be removed.

RezwanArefin01 commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -22,6 +24,7 @@ public class ModelManager implements Model {
     private final AddressBook addressBook;
     private final UserPrefs userPrefs;
     private final FilteredList<Person> filteredPersons;
+    private final FocusMode focusMode;
```
Is a `FocusMode` class really necessary? It is essentially storing a `Tag` (later `Class`) object. 
The model class already has many functions that interact with the intermediate objects of the system. So wouldn't it be better to store a `Tag focusedClass` instead of the current `FocusMode` here, and move the functionality of `FocusMode` inside `ModelManager`?

RezwanArefin01 commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -147,4 +151,19 @@ public boolean equals(Object obj) {
+    @Override
+    public void enterFocusMode(Tag classToFocus) {
+        requireNonNull(classToFocus);
+        focusMode.enter(classToFocus);
```
Following the above suggestion, this would be `focusedClass = classToFocus` instead.

RezwanArefin01 reviewed (changes requested):
Suggested some abstraction changes. See what you think about them.

njxue committed:
Merge branch 'master' of https://github.com/AY2223S1-CS2103T-T12-1/tp into track-focus-mode

njxue committed:
Move functionality of focus mode from FocusMode to ModelManager

FocusMode class is unncessary because it is essentially just storing
the ModuleClass object.

Let's let ModelManager handle the state of the focus mode instead.

njxue committed:
Fix style issues

njxue committed:
Add period to class header comment

njxue commented:
Ignore variable names and comments with "tag" in this PR, cos @Bubbl3T will fix it later

Bubbl3T reviewed (approved):
LGTM!

Bubbl3T committed:
Merge branch 'master' into track-focus-mode

Bubbl3T committed:
Change person to student

njxue committed:
Resolve merge conflicts

njxue committed:
Resolve merge conflicts

njxue committed:
Remove blank line

Bubbl3T commented:
Yeap, looks good to merge! :) Thanks for the great work!
